python_modules/dagster/dagster/_core/definitions/metadata/source_code.py

```diff
@@ -130,6 +143,94 @@ def _with_code_source_single_definition(
     return assets_def.with_attributes(metadata_by_key=metadata_by_key)
 
 
+def convert_local_path_to_source_control_path(
+    base_source_control_url: str,
+    repository_root_absolute_path: str,
+    local_path: LocalFileCodeReference,
+) -> SourceControlCodeReference:
+    source_file_from_repo_root = os.path.relpath(
+        local_path.file_path, repository_root_absolute_path
+    )
+
+    return SourceControlCodeReference(
+        source_control_url=f"{base_source_control_url}/{source_file_from_repo_root}",
+        line_number=local_path.line_number,
+        label=local_path.label,
+    )
+
+
+def _convert_local_path_to_source_control_path_single_definition(
+    base_source_control_url: str,
+    repository_root_absolute_path: str,
+    assets_def: Union["AssetsDefinition", "SourceAsset", "CacheableAssetsDefinition"],
+) -> Union["AssetsDefinition", "SourceAsset", "CacheableAssetsDefinition"]:
+    from dagster._core.definitions.assets import AssetsDefinition
+
+    # SourceAsset doesn't have an op definition to point to - cacheable assets
+    # will be supported eventually but are a bit trickier
+    if not isinstance(assets_def, AssetsDefinition):
+        return assets_def
+
+    metadata_by_key = dict(assets_def.metadata_by_key) or {}
+
+    for key in assets_def.keys:
+        try:
+            existing_source_code_metadata = CodeReferencesMetadataSet.extract(
+                metadata_by_key.get(key, {})
+            )
+            sources_for_asset: List[Union[LocalFileCodeReference, SourceControlCodeReference]] = [
+                convert_local_path_to_source_control_path(
+                    base_source_control_url,
+                    repository_root_absolute_path,
+                    source,
+                )
+                if isinstance(source, LocalFileCodeReference)
+                else source
+                for source in existing_source_code_metadata.code_references.code_references
+            ]
+            metadata_by_key[key] = {
+                **metadata_by_key.get(key, {}),
+                **CodeReferencesMetadataSet(
+                    code_references=CodeReferencesMetadataValue(code_references=sources_for_asset)
+                ),
+            }
+        except pydantic.ValidationError:
```

Comment:
Why ignore error here?

Reply:
In case `CodeReferencesMetadataSet.extract` fails, e.g. the metadata with the corresponding key isn't in the expected format. We could instead loudly error here - the only case where I would really expect this to happen is if the user manually defined one of the code references keys but with their own value type.

Reply:
I think we should loudly error instead so they know to fix it

Reply:
Actually, at second glance, `extract` will fail if any of the required keys don't exist in the metadata - so this `ValidationError` triggers on any asset w/o any code references metadata, too.